### DISCUSSION TITLE
Is there a way of replacing the logging formatter?

### DISCUSSION THREAD

**Discussion by EpicBirb** [abhinavsingh/proxy.py#1289] [Q&A] [answered]
As the title says

Refer to this: https://docs.python.org/3/library/logging.html#logging.Formatter

**Comment by abhinavsingh** [comment, marked as answer]
@EpicBirb Yes. You can either:

1) Pass custom `--log-format ` argument to `proxy` instance
2) Write a `on_access_log` handler in your plugin which does all the logging for you.   Your handler can return None as context which will disable core logging.

**Reply by EpicBirb** [reply]
for the second solution. is it possible to fetch if the context from `on_access_log` is either an info, error, critical, warn, or debug?

because all i really want to do is modify `2022-10-31 16:47:50,632 - pid:26428 [I] pool.shutdown:125 - Shutting down 16 acceptors` to have some color so that my eyes would feel better to identify if its an error or an info because its a huge pain

**Comment by abhinavsingh** [comment]
Of-course, `[I]` you see in logs is the log level.

I = INFO

**Reply by EpicBirb** [reply]
ok here let me reword this:

is there any way that i can modify the logs:
example:
```
2022-11-01 15:57:27,227 - pid:15424 [I] handler.handle_data:179 - HttpProtocolException: ProxyAuthenticationFailed

to

\n 2022-11-01 15:57:27,227 - pid:15424 [I] handler.handle_data:179 - HttpProtocolException: ProxyAuthenticationFailed

depending on what it is logging like info, error, etc
i can modify it based on that

\s 2022-11-01 15:57:27,227 - pid:15424 [E] handler.handle_data:179 - HttpProtocolException: ProxyAuthenticationFailed
\n 2022-11-01 15:57:27,227 - pid:15424 [I] handler.handle_data:179 - HttpProtocolException: ProxyAuthenticationFailed
\r 2022-11-01 15:57:27,227 - pid:15424 [D] handler.handle_data:179 - HttpProtocolException: ProxyAuthenticationFailed

**Comment by abhinavsingh** [comment]
As I explained it's very much possible.  You can take inspiration from ProgramNamePlugin which adds updates `client_ip` with detected program name in the context.  https://github.com/abhinavsingh/proxy.py/blob/develop/proxy/plugin/program_name.py#L64

You can add your own keys too.  Those keys are then available within your log format.

You can also return None from your `on_access_log` callback which will disable internal logging.  You can then log whatever you want using the information in context variable or elsewhere in your program.

Best

**Reply by EpicBirb** [reply]
but another issue arises,
```
2022-11-01 18:00:17,409 - pid:25740 [E] handler.run:341 - ssl.SSLError
Traceback (most recent call last):
  File "e:\projects\test\proxy.py\proxy\http\handler.py", line 327, in run
    self.initialize()
  File "e:\projects\test\proxy.py\proxy\http\handler.py", line 63, in initialize
    super().initialize()
  File "e:\projects\test\proxy.py\proxy\core\base\tcp_server.py", line 129, in initialize
    conn = self._optionally_wrap_socket(self.work.connection)
  File "e:\projects\test\proxy.py\proxy\core\base\tcp_server.py", line 240, in _optionally_wrap_socket
    conn = wrap_socket(conn, self.flags.keyfile, self.flags.certfile)
  File "e:\projects\test\proxy.py\proxy\common\utils.py", line 232, in wrap_socket
    return ctx.wrap_socket(
  File "C:\Users\re\AppData\Local\Programs\Python\Python39\lib\ssl.py", line 500, in wrap_socket
    return self.sslsocket_class._create(
  File "C:\Users\re\AppData\Local\Programs\Python\Python39\lib\ssl.py", line 1040, in _create
    self.do_handshake()
  File "C:\Users\re\AppData\Local\Programs\Python\Python39\lib\ssl.py", line 1309, in do_handshake
    self._sslobj.do_handshake()
ssl.SSLError: [SSL: HTTP_REQUEST] http request (_ssl.c:1122)
```
![image](https://user-images.githubusercontent.com/68750348/199350670-d9acc0b0-c957-4b96-ad2f-19a667cac9a8.png)

it doesn't exactly pass in errors too.

**Reply by abhinavsingh** [reply]
Please check https://github.com/abhinavsingh/proxy.py#oserror-when-wrapping-client-for-tls-interception

**Reply by abhinavsingh** [reply]
You must also provide more context to help reproduce the error.  E.g. what exactly was tried.

**Reply by EpicBirb** [reply]
this isint exactly an issue, i just need to make an error to say that errors wont be passed though on_access_log

if you still want to reproduce this, enable end to end encryption and connect to the proxy using http (without the s)

**Reply by abhinavsingh** [reply]
No, access_log is not an exception handler :).  No traceback or errors are passed to it currently.  Thank you for confirming that it works as expected

**Reply by EpicBirb** [reply]
welp that sucks
